src/main/java/org/opensearch/plugin/insights/core/service/categorizer/QueryShapeGenerator.java

```diff
@@ -195,11 +225,60 @@ static String buildSortShape(List<SortBuilder<?>> sortBuilderList, Boolean showF
      * @return String: comma separated list with leading space in square brackets
      * Ex: " [my_field, width:5]"
      */
-    static String buildFieldDataString(NamedWriteable builder) {
+    static String buildFieldDataString(NamedWriteable builder, Metadata metadata, Set<String> searchIndices) {
         List<String> fieldDataList = new ArrayList<>();
         if (builder instanceof WithFieldName) {
-            fieldDataList.add(((WithFieldName) builder).fieldName());
+            String fieldName = ((WithFieldName) builder).fieldName();
+            fieldDataList.add(fieldName);
+            String fieldType = getFieldType(fieldName, metadata, searchIndices);
+            if (fieldType != null) {
+                fieldDataList.add(fieldType);
+            }
         }
         return " [" + String.join(", ", fieldDataList) + "]";
     }
+
+    /**
+     * Method to get a field's type
+     * @return String field type
+     */
+    static String getFieldType(String fieldName, Metadata metadata, Set<String> searchIndices) {
+        if (metadata == null) {
+            return null;
+        }
+        for (String searchIndex : searchIndices) {
+            IndexMetadata indexMetadata = metadata.index(searchIndex);
+            if (indexMetadata == null) {
+                continue;
+            }
+            MappingMetadata mappingMetadata = indexMetadata.mapping();
+            if (mappingMetadata == null) {
+                continue;
+            }
+            Map<String, Object> sourceMap = mappingMetadata.getSourceAsMap();
+
+            String fieldType = findFieldType((Map<String, Object>) sourceMap.get("properties"), fieldName);
+            if (fieldType != null) {
+                return fieldType;
+            }
+        }
+        return null;
+    }
+
+    public static String findFieldType(Map<String, Object> properties, String fieldName) {
+        for (Map.Entry<String, Object> entry : properties.entrySet()) {
+            Map<String, Object> field = (Map<String, Object>) entry.getValue();
+            if (entry.getKey().equals(fieldName)) {
+                return (String) field.get("type");
+            }
+            // TODO: Add support for multi-fields
```

Comment:
When can we add this?

Reply:
This will need to come in a future iteration. The challenge is that a single multi field can be queried with `name`, `name.keyword`, `name.lowercase`, `name.initials` with the example mapping below. Also non-multi fields can contain "." in the field name. So `name.initials` could be a multi field or just a normal field name.
```
{
  "mappings": {
    "properties": {
      "name": {
        "type": "text",
        "fields": {
          "keyword": {
            "type": "keyword",
            "ignore_above": 256
          },
          "lowercase": {
            "type": "text",
            "normalizer": "lowercase_normalizer"
          },
          "initials": {
            "type": "text",
            "fields": {
              "keyword": {
                "type": "keyword",
                "ignore_above": 256
              }
            }
          }
        }
      }
    }
  },
  "settings": {
    "analysis": {
      "normalizer": {
        "lowercase_normalizer": {
          "type": "custom",
          "filter": ["lowercase"]
        }
      }
    }
  }
}
```